// panda-river/build/npm/src/reactor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isReactor = exports.reactor = exports.isReagent = void 0;

var _pandaGarden = require("panda-garden");

var _pandaGenerics = require("panda-generics");

var _pandaParchment = require("panda-parchment");

function _awaitAsyncGenerator(value) { return new _AwaitValue(value); }

function _wrapAsyncGenerator(fn) { return function () { return new _AsyncGenerator(fn.apply(this, arguments)); }; }

function _AsyncGenerator(gen) { var front, back; function send(key, arg) { return new Promise(function (resolve, reject) { var request = { key: key, arg: arg, resolve: resolve, reject: reject, next: null }; if (back) { back = back.next = request; } else { front = back = request; resume(key, arg); } }); } function resume(key, arg) { try { var result = gen[key](arg); var value = result.value; var wrappedAwait = value instanceof _AwaitValue; Promise.resolve(wrappedAwait ? value.wrapped : value).then(function (arg) { if (wrappedAwait) { resume("next", arg); return; } settle(result.done ? "return" : "normal", arg); }, function (err) { resume("throw", err); }); } catch (err) { settle("throw", err); } } function settle(type, value) { switch (type) { case "return": front.resolve({ value: value, done: true }); break; case "throw": front.reject(value); break; default: front.resolve({ value: value, done: false }); break; } front = front.next; if (front) { resume(front.key, front.arg); } else { back = null; } } this._invoke = send; if (typeof gen.return !== "function") { this.return = undefined; } }

if (typeof Symbol === "function" && Symbol.asyncIterator) { _AsyncGenerator.prototype[Symbol.asyncIterator] = function () { return this; }; }

_AsyncGenerator.prototype.next = function (arg) { return this._invoke("next", arg); };

_AsyncGenerator.prototype.throw = function (arg) { return this._invoke("throw", arg); };

_AsyncGenerator.prototype.return = function (arg) { return this._invoke("return", arg); };

function _AwaitValue(value) { this.wrapped = value; }

var asyncIterator, isAsyncGeneratorFunction, isAsyncIterable, isAsyncIterator, isReactor, isReagent, reactor;
exports.reactor = reactor;
exports.isReagent = isReagent;
exports.isReactor = isReactor;

if (Symbol.asyncIterator == null) {
  Symbol.asyncIterator = Symbol("asyncIterator");
}

isAsyncGeneratorFunction = function () {
  var f;

  f =
  /*#__PURE__*/
  function () {
    var _ref = _wrapAsyncGenerator(function* () {
      return yield null;
    });

    return function f() {
      return _ref.apply(this, arguments);
    };
  }();

  return (0, _pandaParchment.isType)(f.constructor);
}();

exports.isReagent = isReagent = isAsyncIterable = function (x) {
  return (0, _pandaParchment.isFunction)(x != null ? x[Symbol.asyncIterator] : void 0) || isAsyncGeneratorFunction(x);
};

exports.isReactor = isReactor = isAsyncIterator = function (x) {
  return (0, _pandaParchment.isFunction)(x != null ? x.next : void 0) && isReagent(x);
};

exports.reactor = reactor = asyncIterator = _pandaGenerics.Method.create({
  default: "unable to create reactor from value"
});

_pandaGenerics.Method.define(reactor, _pandaParchment.isFunction, function (f) {
  return {
    next: f,
    [Symbol.asyncIterator]: function () {
      return this;
    }
  };
});

_pandaGenerics.Method.define(reactor, isReagent, function (r) {
  return r[Symbol.asyncIterator]();
});

_pandaGenerics.Method.define(reactor, isAsyncGeneratorFunction, function (g) {
  return g();
});

_pandaGenerics.Method.define(reactor, isReactor, function (r) {
  return r;
});
