// panda-template/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _fairmont = require("fairmont");

var _sprintfJs = require("sprintf-js");

var _sprintfJs2 = _interopRequireDefault(_sprintfJs);

var _handlebars = require("handlebars");

var _handlebars2 = _interopRequireDefault(_handlebars);

var _swag = require("swag");

var _swag2 = _interopRequireDefault(_swag);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Generated by CoffeeScript 2.0.2
var PandaTemplate,
    block,
    join,
    slice = [].slice;

_swag2.default.registerHelpers(_handlebars2.default);

join = function (d, array) {
  return array.join(d);
};

block = function (f) {
  return function (...args) {
    var j, options, ref;
    ref = args, args = 2 <= ref.length ? slice.call(ref, 0, j = ref.length - 1) : (j = 0, []), options = ref[j++];
    if (options.fn != null) {
      return join("", f(...args, options.fn));
    } else {
      return f(...args, _fairmont.identity);
    }
  };
};

_handlebars2.default.registerHelper({
  let: function (...args) {
    var j, options, ref;
    ref = args, args = 2 <= ref.length ? slice.call(ref, 0, j = ref.length - 1) : (j = 0, []), options = ref[j++];
    return options.fn(this, {
      data: options.data,
      blockParams: args
    });
  },
  values: block(function (object, f) {
    var key, results, value;
    results = [];
    for (key in object) {
      value = object[key];
      results.push(f(value));
    }
    return results;
  }),
  filter: block(function (property, value, objects, f) {
    var j, len, object, results;
    results = [];
    for (j = 0, len = objects.length; j < len; j++) {
      object = objects[j];
      if (object[property] === value) {
        results.push(f(object));
      }
    }
    return results;
  }),
  pluck: block(function (property, objects, f) {
    var j, len, object, results;
    results = [];
    for (j = 0, len = objects.length; j < len; j++) {
      object = objects[j];
      results.push(f(object[property]));
    }
    return results;
  }),
  join: block(function (delimiter, array, f) {
    return [join(delimiter, array.map(f))];
  }),
  sprintf: function (format, string) {
    return (0, _sprintfJs2.default)(format, string);
  },
  indent: function (n, string) {
    var indent;
    indent = " ".repeat(n);
    return string.replace(/\n/g, `\n${indent}`);
  },
  hang: function () {
    var indent;
    indent = function (x, s) {
      return " ".repeat(x) + s;
    };
    return function (i, w, s) {
      var first, line, m, rest;
      m = w - i;
      [first, ...rest] = s.match(RegExp(`.{1,${m}}(\\s+|$)`, "g"));
      return [first, ...function () {
        var j, len, results;
        results = [];
        for (j = 0, len = rest.length; j < len; j++) {
          line = rest[j];
          results.push(indent(i, line));
        }
        return results;
      }()].join("\n");
    };
  }(),
  json: function (value) {
    return JSON.stringify(value, null, 2);
  },
  match: block(function (re, s, f) {
    var j, k, len, results, v;
    if (k = s.match(new RegExp(re, "g"))) {
      results = [];
      for (j = 0, len = k.length; j < len; j++) {
        v = k[j];
        results.push(f(v));
      }
      return results;
    } else {
      return [];
    }
  }),
  replace: function (re, r, s) {
    return s.replace(new RegExp(re, "g"), r);
  },
  pick: function (...args) {
    return args.find(_fairmont.isDefined);
  },
  parsePath: block(function (s, f) {
    var j, len, p, ref, results;
    ref = ["/", ...s.split("/").slice(1)];
    results = [];
    for (j = 0, len = ref.length; j < len; j++) {
      p = ref[j];
      results.push(f(p));
    }
    return results;
  })
});

PandaTemplate = class PandaTemplate {
  constructor() {
    this.H = _handlebars2.default;
  }

  render(template, context) {
    return this.H.compile(template, {
      noEscape: true
    })(context);
  }

  registerPartial(name, template) {
    return this.H.registerPartial(name, template);
  }

  handlebars() {
    return this.H;
  }

};

exports.default = PandaTemplate;
