// sundog/build/npm/src/primitives/s3/object/base.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _pandaParchment = require("panda-parchment");

var _privateUtils = require("../../private-utils");

var Section;

Section = function (s3) {
  var deleteBatch, destroy, exists, get, head, list;

  head = exists = async function (name, key) {
    var e;

    try {
      return await s3.headObject({
        Bucket: name,
        Key: key
      });
    } catch (error) {
      e = error;
      return (0, _privateUtils.notFound)(e);
    }
  };

  get = async function (name, key, encoding = "utf8") {
    var Body, e;

    try {
      ({
        Body
      } = await s3.getObject({
        Bucket: name,
        Key: key
      }));

      if (encoding === "binary") {
        return Body;
      } else {
        return Body.toString(encoding);
      }
    } catch (error) {
      e = error;
      return (0, _privateUtils.notFound)(e);
    }
  };

  destroy = async function (name, key) {
    var e;

    try {
      return await s3.deleteObject({
        Bucket: name,
        Key: key
      });
    } catch (error) {
      e = error;
      return (0, _privateUtils.notFound)(e);
    }
  };

  deleteBatch = async function (name, keys) {
    var key;
    return await s3.deleteObjects({
      Bucket: name,
      Delete: {
        Objects: function () {
          var i, len, results;
          results = [];

          for (i = 0, len = keys.length; i < len; i++) {
            key = keys[i];
            results.push({
              Key: key
            });
          }

          return results;
        }(),
        Quiet: true
      }
    });
  };

  list = async function (name, items = [], marker) {
    var Contents, IsTruncated, NextContinuationToken, p;
    p = {
      Bucket: name,
      MaxKeys: 1000
    };

    if (marker) {
      p.ContinuationToken = marker;
    }

    ({
      IsTruncated,
      Contents,
      NextContinuationToken
    } = await s3.listObjectsV2(p));

    if (IsTruncated) {
      items = (0, _pandaParchment.cat)(items, Contents);
      return await list(name, items, NextContinuationToken);
    } else {
      return (0, _pandaParchment.cat)(items, Contents);
    }
  };

  return {
    head,
    exists,
    get,
    delete: destroy,
    deleteBatch,
    list
  };
};

var _default = Section;
exports.default = _default;
