// panda-river/build/npm/test/reducers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _assert = _interopRequireDefault(require("assert"));

var _amen = require("amen");

var _pandaGarden = require("panda-garden");

var _pandaParchment = require("panda-parchment");

var _reducers = require("../src/reducers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _awaitAsyncGenerator(value) { return new _AwaitValue(value); }

function _wrapAsyncGenerator(fn) { return function () { return new _AsyncGenerator(fn.apply(this, arguments)); }; }

function _AsyncGenerator(gen) { var front, back; function send(key, arg) { return new Promise(function (resolve, reject) { var request = { key: key, arg: arg, resolve: resolve, reject: reject, next: null }; if (back) { back = back.next = request; } else { front = back = request; resume(key, arg); } }); } function resume(key, arg) { try { var result = gen[key](arg); var value = result.value; var wrappedAwait = value instanceof _AwaitValue; Promise.resolve(wrappedAwait ? value.wrapped : value).then(function (arg) { if (wrappedAwait) { resume("next", arg); return; } settle(result.done ? "return" : "normal", arg); }, function (err) { resume("throw", err); }); } catch (err) { settle("throw", err); } } function settle(type, value) { switch (type) { case "return": front.resolve({ value: value, done: true }); break; case "throw": front.reject(value); break; default: front.resolve({ value: value, done: false }); break; } front = front.next; if (front) { resume(front.key, front.arg); } else { back = null; } } this._invoke = send; if (typeof gen.return !== "function") { this.return = undefined; } }

if (typeof Symbol === "function" && Symbol.asyncIterator) { _AsyncGenerator.prototype[Symbol.asyncIterator] = function () { return this; }; }

_AsyncGenerator.prototype.next = function (arg) { return this._invoke("next", arg); };

_AsyncGenerator.prototype.throw = function (arg) { return this._invoke("throw", arg); };

_AsyncGenerator.prototype.return = function (arg) { return this._invoke("return", arg); };

function _AwaitValue(value) { this.wrapped = value; }

function _asyncIterator(iterable) { var method; if (typeof Symbol === "function") { if (Symbol.asyncIterator) { method = iterable[Symbol.asyncIterator]; if (method != null) return method.call(iterable); } if (Symbol.iterator) { method = iterable[Symbol.iterator]; if (method != null) return method.call(iterable); } } throw new TypeError("Object is not async iterable"); }

var spec;

spec = function (name, {
  expected,
  reducer,
  iterable
}) {
  return (0, _amen.test)(name, [(0, _amen.test)("iterator", function () {
    return _assert.default.deepEqual(expected, reducer(iterable));
  }), (0, _amen.test)("reactor", async function () {
    var r;

    r =
    /*#__PURE__*/
    function () {
      var _ref = _wrapAsyncGenerator(function* () {
        var results, x;
        results = [];
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;

        var _iteratorError;

        try {
          for (var _iterator = _asyncIterator(iterable), _step, _value; _step = yield _awaitAsyncGenerator(_iterator.next()), _iteratorNormalCompletion = _step.done, _value = yield _awaitAsyncGenerator(_step.value), !_iteratorNormalCompletion; _iteratorNormalCompletion = true) {
            x = _value;
            results.push((yield x));
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator.return != null) {
              yield _awaitAsyncGenerator(_iterator.return());
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }

        return results;
      });

      return function r() {
        return _ref.apply(this, arguments);
      };
    }();

    return _assert.default.deepEqual(expected, (await reducer(r)));
  })]);
};

var _default = [spec("collect", {
  iterable: [1, 2, 3, 4, 5],
  expected: [1, 2, 3, 4, 5],
  reducer: _reducers.collect
}), spec("each", {
  iterable: [1, 2, 3, 4, 5],
  expected: void 0,
  reducer: (0, _reducers.each)(function (y) {
    return function (x) {
      (0, _assert.default)(x === y);
      return y = y % 5 + 1;
    };
  }(1))
}), spec("fold/reduce", {
  iterable: [1, 2, 3, 4, 5],
  expected: 15,
  reducer: (0, _reducers.fold)(_pandaParchment.add, 0)
}), spec("foldr/reduceRight", {
  iterable: "panama",
  expected: "amanap",
  reducer: (0, _reducers.foldr)(_pandaParchment.add, "")
}), spec("any", {
  iterable: [1, 2, 3, 4, 5],
  expected: true,
  reducer: (0, _reducers.any)(_pandaParchment.odd)
}), spec("all", {
  iterable: [1, 2, 3, 4, 5],
  expected: false,
  reducer: (0, _reducers.all)(_pandaParchment.odd)
}), spec("sum", {
  iterable: [1, 2, 3, 4, 5],
  expected: 15,
  reducer: _reducers.sum
}), spec("average", {
  iterable: [1, 2, 3, 4, 5],
  expected: 3,
  reducer: _reducers.average
}), spec("delimit", {
  iterable: ["one", "two", "three"],
  expected: "one, two, three",
  reducer: (0, _reducers.delimit)(", ")
})];
exports.default = _default;
