// sundog/build/npm/src/primitives/cloudwatchlogs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _pandaParchment = require("panda-parchment");

var _lift = require("../lift");

// Primitives for the service CloudWatch Logs.  The main entities are groups and streams.
// This follows the naming convention that methods that work on groups will be
// prefixed "group*", whereas streams methods will have no prefix.
var cloudwatchPrimitive;

cloudwatchPrimitive = function (SDK) {
  return function (configuration) {
    var create, exists, groupList, latest, logs, tail;
    logs = (0, _lift.applyConfiguration)(configuration, SDK.CloudWatchLogs); // Returns data on a group or groups given an input prefix.

    groupList = async function (prefix, current = [], token) {
      var logGroups, nextToken, params;
      params = {
        logGroupNamePrefix: prefix
      };

      if (token) {
        params.nextToken = token;
      }

      ({
        logGroups,
        nextToken
      } = await logs.describeLogGroups(params));
      current = (0, _pandaParchment.cat)(current, logGroups);

      if (nextToken) {
        return await groupList(prefix, current, nextToken);
      } else {
        return current;
      }
    };

    latest = async function (name) {
      var logStreams, params;
      params = {
        logGroupName: name,
        orderBy: "LastEventTime",
        descending: true,
        limit: 1
      };
      ({
        logStreams
      } = await logs.describeLogStreams(params));

      if ((0, _pandaParchment.empty)(logStreams)) {
        return void 0;
      } else {
        return logStreams[0];
      }
    };

    tail = async function (group, stream, time, current = [], token = false) {
      var events, nextForwardToken, params;
      params = {
        logGroupName: group,
        logStreamName: stream,
        startTime: time,
        startFromHead: true
      };

      if (token) {
        params.nextToken = token;
      }

      ({
        events,
        nextForwardToken
      } = await logs.getLogEvents(params));
      current = (0, _pandaParchment.cat)(current, events);

      if (nextForwardToken !== token) {
        return await tail(group, stream, time, current, nextForwardToken);
      } else {
        return current;
      }
    };

    exists = async function (name) {
      var logGroups;
      ({
        logGroups
      } = await logs.describeLogGroups({
        limit: 1,
        logGroupNamePrefix: name
      }));

      if ((0, _pandaParchment.empty)(logGroups)) {
        return false;
      } else {
        return true;
      }
    };

    create = async function (name) {
      return await logs.createLogGroup({
        logGroupName: name
      });
    }; // Return exposed functions.


    return {
      groupList,
      latest,
      tail,
      exists,
      create
    };
  };
};

var _default = cloudwatchPrimitive;
exports.default = _default;
