// sundog/build/npm/src/primitives/cloudfront.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _pandaParchment = require("panda-parchment");

var _pandaRiver = require("panda-river");

var _privateUtils = require("./private-utils");

var _url = require("../helpers/url");

var _kms = _interopRequireDefault(require("./kms"));

var _lift = require("../lift");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var cloudfrontPrimitive;

cloudfrontPrimitive = function (SDK) {
  return function (configuration) {
    var cfr, get, hoistETag, invalidate, list, randomBytes;
    cfr = (0, _lift.applyConfiguration)(configuration, SDK.CloudFront);
    ({
      randomBytes
    } = (0, _kms.default)(SDK)(configuration));

    list = async function (current = [], marker) {
      var data, params;
      params = {
        MaxItems: "100"
      };

      if (marker) {
        params.Marker = marker;
      }

      data = await cfr.listDistributions(params);
      current = (0, _pandaParchment.cat)(current, data.DistributionList.Items);

      if (data.IsTruncated) {
        return await list(current, data.DistributionList.Marker);
      } else {
        return current;
      }
    };

    hoistETag = function (data) {
      var Distribution;
      ({
        Distribution
      } = data);
      return (0, _pandaParchment.merge)(Distribution, {
        ETag: data.ETag
      });
    };

    get = async function (name) {
      var matches, pattern;
      list = await list();
      pattern = {
        Aliases: {
          Quantity: 1,
          Items: [(0, _url.regularlyQualify)(name)]
        }
      };
      matches = (0, _pandaRiver.collect)((0, _privateUtils.where)(pattern, list));

      if ((0, _pandaParchment.empty)(matches)) {
        return false;
      } else {
        return hoistETag((await cfr.getDistribution({
          Id: matches[0].Id
        })));
      }
    };

    invalidate = async function ({
      Id
    }, paths) {
      var Invalidation, Status, params;

      if (paths == null) {
        paths = ["/*"];
      }

      params = {
        DistributionId: Id,
        InvalidationBatch: {
          CallerReference: "Sky" + (await randomBytes(16)).toString(),
          Paths: {
            Quantity: paths.length,
            Items: paths
          }
        }
      };
      ({
        Invalidation
      } = await cfr.createInvalidation(params));
      params = {
        DistributionId: Id,
        Id: Invalidation.Id
      };

      while (true) {
        ({
          Invalidation: {
            Status
          }
        } = await cfr.getInvalidation(params));

        if (Status === "Completed") {
          return;
        } else {
          await (0, _pandaParchment.sleep)(15000);
        }
      }
    };

    return {
      list,
      get,
      invalidate
    };
  };
};

var _default = cloudfrontPrimitive;
exports.default = _default;
