// sundog/build/npm/src/primitives/lambda.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _pandaParchment = require("panda-parchment");

var _lift = require("../lift");

var lambdaPrimitive;

lambdaPrimitive = function (SDK) {
  return function (configuration) {
    var Delete, asyncInvoke, invoke, lambda, list, update, updateConfig;
    lambda = (0, _lift.applyConfiguration)(configuration, SDK.Lambda);

    update = async function (name, bucket, key, options = {}) {
      return await lambda.updateFunctionCode((0, _pandaParchment.merge)({
        FunctionName: name,
        Publish: false,
        S3Bucket: bucket,
        S3Key: key
      }, options));
    };

    updateConfig = async function (FunctionName, config) {
      return await lambda.updateFunctionConfiguration((0, _pandaParchment.merge)({
        FunctionName
      }, config));
    };

    list = async function (fns = [], marker) {
      var Functions, NextMarker, params;
      params = {
        MaxItems: 100
      };

      if (marker) {
        params.Marker = marker;
      }

      ({
        NextMarker,
        Functions
      } = await lambda.listFunctions(params));
      fns = (0, _pandaParchment.cat)(fns, Functions);

      if (NextMarker) {
        return await list(fns, NextMarker);
      } else {
        return fns;
      }
    };

    Delete = async function (name) {
      return await lambda.deleteFunction({
        FunctionName: name
      });
    };

    invoke = async function (FunctionName, input, options) {
      var params;
      params = {
        FunctionName,
        Payload: JSON.stringify(input)
      };
      params = (0, _pandaParchment.merge)(params, options);
      return await lambda.invoke(params);
    };

    asyncInvoke = function (name, input, options) {
      options = (0, _pandaParchment.merge)(options, {
        InvocationType: "Event"
      });
      return invoke(name, input, options);
    };

    return {
      update,
      updateConfig,
      list,
      delete: Delete,
      invoke,
      asyncInvoke
    };
  };
};

var _default = lambdaPrimitive;
exports.default = _default;
