// sundog/build/npm/src/primitives/step.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _pandaParchment = require("panda-parchment");

var _lift = require("../lift");

// Primitives for the service StepFunctions.
var stepPrimitive;

stepPrimitive = function (SDK) {
  return function (configuration) {
    var list, listExes, lookup, start, step, stop, stopAll;
    step = (0, _lift.applyConfiguration)(configuration, SDK.StepFunctions);

    start = function (arn, input, name) {
      return step.startExecution({
        stateMachineArn: arn,
        input: input ? (0, _pandaParchment.toJSON)(input) : void 0,
        name: name
      });
    };

    stop = function (arn, options) {
      return step.stopExecution((0, _pandaParchment.merge)({
        executionArn: arn
      }, options));
    };

    listExes = async function (arn, status, current = [], next) {
      var executions, nextToken;
      ({
        nextToken,
        executions
      } = await step.listExecutions({
        stateMachineArn: arn,
        statusFilter: status,
        maxResults: 1000,
        nextToken: next
      }));
      current = (0, _pandaParchment.cat)(current, executions);

      if (nextToken) {
        return listExes(arn, status, current, nextToken);
      } else {
        return current;
      }
    };

    list = async function (current = [], next) {
      var nextToken, stateMachines;
      ({
        nextToken,
        stateMachines
      } = await step.listStateMachines({
        maxResults: 1000,
        nextToken: next
      }));
      current = (0, _pandaParchment.cat)(current, stateMachines);

      if (nextToken) {
        return list(current, nextToken);
      } else {
        return current;
      }
    };

    lookup = async function (name) {
      var i, len, machine, machines;
      machines = await list();

      for (i = 0, len = machines.length; i < len; i++) {
        machine = machines[i];

        if (machine.name === name) {
          return machine;
        }
      }

      return false;
    };

    stopAll = async function (name) {
      var executionArn, expressions, i, len, results, stateMachineArn;
      ({
        stateMachineArn
      } = await lookup(name));
      expressions = await listExes(stateMachineArn, "RUNNING");
      results = [];

      for (i = 0, len = expressions.length; i < len; i++) {
        ({
          executionArn
        } = expressions[i]);
        results.push((await stop(executionArn)));
      }

      return results;
    };

    return {
      start,
      stop,
      list,
      lookup,
      listExes,
      stopAll
    };
  };
};

var _default = stepPrimitive;
exports.default = _default;
