// fairmont-helpers/lib/numeric.js
// Generated by CoffeeScript 1.12.4
(function() {
  var abs, add, curry, div, eq, even, gt, gte, lt, lte, max, min, mod, mul, negate, neq, odd, partial, pow, ref, sub;

  ref = require("fairmont-core"), curry = ref.curry, partial = ref.partial, negate = ref.negate;

  eq = curry(function(x, y) {
    return x === y;
  });

  neq = curry(function(x, y) {
    return x !== y;
  });

  gte = curry(function(x, y) {
    return y >= x;
  });

  lte = curry(function(x, y) {
    return y <= x;
  });

  gt = curry(function(x, y) {
    return y > x;
  });

  lt = curry(function(x, y) {
    return y < x;
  });

  add = curry(function(x, y) {
    return x + y;
  });

  sub = curry(function(x, y) {
    return y - x;
  });

  mul = curry(function(x, y) {
    return x * y;
  });

  div = curry(function(x, y) {
    return y / x;
  });

  mod = curry(function(x, y) {
    return y % x === 0;
  });

  even = mod(2);

  odd = negate(even);

  min = Math.min, max = Math.max, abs = Math.abs, pow = Math.pow;

  module.exports = {
    eq: eq,
    neq: neq,
    gt: gt,
    lt: lt,
    gte: gte,
    lte: lte,
    add: add,
    sub: sub,
    mul: mul,
    div: div,
    mod: mod,
    even: even,
    odd: odd,
    min: min,
    max: max,
    abs: abs
  };

}).call(this);
