// sundog/build/npm/src/primitives/ec2/eni.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _pandaParchment = require("panda-parchment");

var _pandaRiver = require("panda-river");

var _privateUtils = require("../private-utils");

var _lift = require("../../lift");

var eniPrimitive;

eniPrimitive = function (SDK) {
  return function (configuration) {
    var Delete, detach, ec2, get, list, purge, waitFor, waitForAvailable, waitForInUse;
    ec2 = (0, _lift.applyConfiguration)(configuration, SDK.EC2);

    list = async function (subnetIDs) {
      var NetworkInterfaces;
      ({
        NetworkInterfaces
      } = await ec2.describeNetworkInterfaces({
        Filters: [{
          Name: "subnet-id",
          Values: subnetIDs
        }]
      }));
      return NetworkInterfaces;
    };

    get = async function (id) {
      var NetworkInterfaces, e;

      try {
        ({
          NetworkInterfaces
        } = await ec2.describeNetworkInterfaces({
          NetworkInterfaceIds: [id]
        }));
        return (0, _pandaParchment.first)(NetworkInterfaces);
      } catch (error) {
        e = error;
        return false;
      }
    };

    waitFor = function (status) {
      return async function (id) {
        var subnet;

        while (true) {
          await (0, _pandaParchment.sleep)(10000);
          subnet = await get(id);

          if (subnet.Status === status) {
            return;
          }
        }
      };
    };

    waitForAvailable = waitFor("available");
    waitForInUse = waitFor("in-use");

    detach = async function (id, attachmentID) {
      await ec2.detachNetworkInterface({
        AttachmentId: attachmentID,
        Force: true
      });
      return await waitForAvailable(id);
    };

    Delete = async function (id) {
      return await ec2.deleteNetworkInterface({
        NetworkInterfaceId: id
      });
    };

    purge = async function (subnetIDs, test) {
      var ENIs, attachedENIs, e; // Collect any ENIs that meet your test specs.

      ENIs = await list(subnetIDs);
      ENIs = (0, _pandaRiver.collect)((0, _pandaRiver.select)(test, ENIs)); // Detach any attached ENIs

      attachedENIs = (0, _pandaRiver.collect)((0, _privateUtils.where)({
        Status: "in-use"
      }, ENIs));
      await Promise.all(function () {
        var i, len, results;
        results = [];

        for (i = 0, len = attachedENIs.length; i < len; i++) {
          e = attachedENIs[i];
          results.push(detach(e.NetworkInterfaceId, e.Attachment.AttachmentId));
        }

        return results;
      }()); // Destroy all the ENIs

      return await Promise.all(function () {
        var i, len, results;
        results = [];

        for (i = 0, len = ENIs.length; i < len; i++) {
          e = ENIs[i];
          results.push(Delete(e.NetworkInterfaceId));
        }

        return results;
      }());
    };

    return {
      purge
    };
  };
};

var _default = eniPrimitive;
exports.default = _default;
